### ISSUE TITLE
cheri128 does not compile to C

### ISSUE THREAD

**Issue by Alasdair** [rems-project/sail#23]
cheri128 does not compile to C. Seems to be an issue with the existential type in the call to HighestSetBits, specifically the well-formedness check that occurs there

Minimal example:

```
default Order dec

$include <flow.sail>
$include <arith.sail>
$include <option.sail>
$include <vector_dec.sail>

val modulus = {ocaml: "modulus", lem: "hardware_mod", coq: "euclid_modulo", _ : "tmod_int"} : forall 'n, 'n > 0 .  (int, atom('n)) -> range(0, 'n - 1)

val add_range = {ocaml: "add_int", lem: "integerAdd", coq: "add_range", c: "add_int"} : forall 'n 'm 'o 'p.
  (range('n, 'm), range('o, 'p)) -> range('n + 'o, 'm + 'p)

val sub_range = {ocaml: "sub_int", lem: "integerMinus", coq: "sub_range"} : forall 'n 'm 'o 'p.
  (range('n, 'm), range('o, 'p)) -> range('n - 'p, 'm - 'o)

val min = {ocaml: "min_int", lem: "min", coq: "min_atom", c:"min_int"} : forall 'a 'b . (atom('a), atom('b)) -> {'c, ('c = 'a | 'c = 'b) & 'c <= 'a & 'c <= 'b . atom('c)}

overload operator + = {add_range}
overload operator - = {sub_range}

infix 1 >>
infix 1 <<

val "shift_bits_right" : forall 'n 'm. (bits('n), bits('m)) -> bits('n) effect {undef}
val "shift_bits_left" : forall 'n 'm. (bits('n), bits('m)) -> bits('n) effect {undef}

val "shiftl" : forall 'm 'n, 'n >= 0. (bits('m), atom('n)) -> bits('m)
val "shiftr" : forall 'm 'n, 'n >= 0. (bits('m), atom('n)) -> bits('m)

overload operator >> = {shift_bits_right, shiftr}
overload operator << = {shift_bits_left, shiftl}

val HighestSetBit : forall 'N , 'N >= 2. bits('N) -> {'n, 0 <= 'n < 'N . (bool, atom('n))}

function HighestSetBit x = {
  foreach (i from ('N - 1) to 0 by 1 in dec)
    if [x[i]] == 0b1 then return (true, i);
  return (false, 0)
}

/* hw rounds up E to multiple of 4 */
function roundUp(e) : range(0, 45) -> range(0, 48) =
    let 'r = modulus(e, 4) in
    if (r == 0)
    then e
    else (e - r + 4)

function computeE (rlength) : bits(65) -> range(0, 48) =
    let (nonzero, 'msb) = HighestSetBit((rlength + (rlength >> 6)) >> 19) in
    if nonzero then
      /* above will always return <= 45 because 19 bits of zero are shifted in from right */
      {assert(0 <= msb & msb <= 45); roundUp (min(msb,45)) }
    else
      0

val main : unit -> unit effect {escape}

function main() = {
  let _ = computeE(0xFFFF_FFFF_FFFF_FFFF @ 0b1);
  ()
}
```

**Comment by rmn30** [comment]
Yes, something is wrong with the C as indicated by the following warning from gcc and increased test failures vs. ocaml:
```
eri128.c: In function ‘zHighestSetBit’:
cheri128.c:72108:50: warning: passing argument 2 of ‘copy_sail_int’ makes pointer from integer without a cast [-Wint-conversion]
           COPY(sail_int)(&((&zgsz33332)->ztup1), zi);
                                                  ^~
In file included from cheri128.c:1:0:
/home/rmn30/scratch/rems/sail2/lib/sail.h:38:8: note: expected ‘const __mpz_struct * {aka const struct <anonymous> *}’ but argument is of type ‘mach_int {aka long int}’
   void copy_ ## type(type *, const type);\
        ^
/home/rmn30/scratch/rems/sail2/lib/sail.h:100:1: note: in expansion of macro ‘SAIL_BUILTIN_TYPE’
 SAIL_BUILTIN_TYPE(sail_int);
 ^~~~~~~~~~~~~~~~~
```
Note that the cheri_cc branch of sail-cheri-mips does not have the HighestSetBit function (it has CountLeadingZeros which is similar) and this doesn't seem to have a problem.

**Comment by Alasdair** [comment]
Should be fixed by 18c49a76, at least I don't get that warning when compiling cheri128 anymore, and the test case I made seems to build correctly.

**Comment by rmn30** [comment]
It works now.
